// nodewidgets/lib/tools/db_oracle.js

var oracle = require("oracle");

var connection = null;
var defaultConnData = { "hostname": "mm-dev-oci", "user": "anhe_ums_head", "password": "om1jda", "database": "ocidev" };



exports.connect = function(connData, onSuccess, onError) {
  connection = null;
  if (connData == null) {
    connData = defaultConnData;
  }
  
  oracle.connect(connData, function(err, conn) {
  
    if (err) {
        console.log(err);
        if (onError) {
          onError(connData, err);
        }
        return;
    }
    connection = conn;
    console.log( "connected" );
    
    if (onSuccess) {
      /*
      conn.execute("SET CLIENT_ENCODING TO 'LATIN1'", [], function(err, results) {
        if (err) {
          console.log(err.toString());
        }
        */
        onSuccess(conn);
      //});
    }
  });
};


exports.getConnection = function() {
  return connection;
};



exports.query = function(strSql, params, fnOk, fnError) {
  exports.getConnection().execute(strSql, params, function(err, results) {
    if ( err ) {
      console.log( "Oracle execute failed: " + strSql );
      console.log( err );
      if (fnError != null) {
        fnError(err);
      }
    } else {
      if (fnOk != null) {
        fnOk(results);
      }
    }
  });
};


exports.queryRow = function(strSql, params, fnOk, fnError) {
  exports.query( strSql, params, function(res) {
    if (res.length > 0) {
      // use 1st result row
      fnOk(res[0]);
    }
    else {
      // give null
      fnOk(null);
    }
  }, fnError);
};


exports.queryValue = function(strSql, params, fnOk, fnError) {
  exports.queryRow( strSql, params, function(res) {
    fnOk(res.RESULT);
  }, fnError);
};



exports.queryPage = function(strSql, offset, limit, params, fnOk, fnError) {
  var strSqlPaged = "SELECT * FROM (" + strSql + ") pagedquery " +
          "WHERE pagedquery.rnum > " + offset + " " +
          " AND pagedquery.rnum <= " + (offset + limit) + " ";
  exports.query( strSqlPaged, params, fnOk, fnError );
};
